recipe/meta.yaml

```diff
@@ -1,113 +1,85 @@
+    # we do not have the correct version. We update keras and would like
+    # to offer at least a working backend: torch.
+    - KERAS_BACKEND=torch
   number: 0
+  script:
+    - {{ PYTHON }} -m pip install . --no-deps --ignore-installed --no-build-isolation -v
+  # skipping s390x since dm-tree and jax are not available
+  skip: true  # [py<39 or s390x]
 
-outputs:
-  - name: keras-base
-    script: install_base.sh  # [not win]
-    script: install_base.bat  # [win]
-    build:
-      noarch: python
-    requirements:
-      host:
-        - python
-        - pip
-        - setuptools
-        - wheel
-      run:
-        - python >=3.6
-        - numpy >=1.9.1
-        - scipy >=0.14
-        - pyyaml
-        - h5py
-        - tensorflow {{ version }}  
-      # add run_constrained to prevent installation of mismatched keras and
-      # keras-base packages
-      run_constrained:
-        - keras {{ version }}
-    test:
-      requires:
-        - pip
-      commands:
-        - pip check
-      imports:
-        - keras
-        - keras.backend
-        - keras.datasets
-        - keras.engine
-        - keras.layers
-        - keras.preprocessing
-        - keras.utils
-        - keras.wrappers
- #some of the imports are commented out during testing to avoid circular dependencies
-  - name: keras
-    script: install_base.sh  # [not win]
-    script: install_base.bat  # [win]
-    build:
-      noarch: python
-    requirements:
-      host:
-        - python
-        - pip
-        - setuptools
-        - wheel
-      run:
-        - python >=3.6
-      run_constrained:
-        - tensorflow {{ version }}
-    test:
-      requires:
-        - pip
-      commands:
-        - pip check
-    about:
-      home: https://github.com/fchollet/keras
-      license: MIT
-      license_family: MIT
-      license_file: LICENSE
-      summary: Deep Learning Library for Theano and TensorFlow
-      description: |
-        Keras is a minimalist, highly modular neural networks library
-        written in Python and capable on running on top of either
-        TensorFlow or Theano. It was developed with a focus on enabling
-        fast experimentation. Being able to go from idea to result with
-        the least possible delay is key to doing good research.
-      doc_url: http://keras.io
-      dev_url: https://github.com/fchollet/keras
+requirements:
+  host:
+    - python
+    - pip
+    - wheel
+    - setuptools
+  run:
+    - python
+    - absl-py
+    - numpy
+    - rich
+    - namex
+    - h5py
+    - dm-tree
+    - ml_dtypes
+    # this is not mentioned in the upstream until version 3.3.0
+    # but it might have been a bug, because it is indeed used
+    # in this version as well at:
+    # https://github.com/keras-team/keras/blob/v3.0.5/keras/utils/torch_utils.py#L3
+    - packaging
+  run_constrained:
+    - tensorflow >=2.15.0,<2.17.0a
+    - jax 0.4.*
+    - pytorch >=2.1,<2.3
```

Comment:
I am a little unsure whether for this specific version, since we do not provide `tensorflow` which was specified as run dependency, we should put `pytorch` in run.

Kind of a judgement call since in theory `keras` can still be installed and imported. Worst case scenario the presence of the backend is checked at runtime and an error (or warning) is thrown. but we did do this with `tensorflow` before.

Reply:
The three supported backends are in `run_constrained`, which is ok, they are optional (if none is installed, `keras` would be useless, but still a valid package).

You are right in mentioning doubts about the versions, which are:

```
jax =0.4.20
torch ~=2.1.0
tensorflow ~=2.16.1
```

which can be found here https://keras.io/getting_started/#tensorflow-compatibility

- we probably used `0.4.*` for `jax` to be less strict about it, and allow the `jax[cuda]` version to be installed (`0.4.24`) -> https://github.com/keras-team/keras/blob/v3.0.5/requirements-jax-cuda.txt .
- `pytorch >=2.1,<2.3` should be `pytorch >=2.1.0,<2.2.0` (expanding the tilde).
- `tensorflow >=2.15.0,<2.17.0a` should be `tensorflow >=2.16.1,<2.17.0` (expanding the tilde).

Reply:
> since we do not provide tensorflow which was specified as run dependency, we should put pytorch in run.

Ah you meant in the previous version.

Good point.

we don't have the right version of tensorflow, and the default backend is still tensorflow (following the latest comments, I restored the default one to `tensorflow`).

Reply:
also on this point, @skupr-anaconda what is your opinion?

Reply:
agree